File: src/main/java/com/example/lee/MinimumWindowSubstring.java
package com.example.lee;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by benbendaisy on 1/14/15.
 * Given a string S and a string T, find the minimum window in S which will contain all the characters in T in complexity O(n).
 *
 * For example,
 * S = "ADOBECODEBANC"
 * T = "ABC"
 * Minimum window is "BANC".
 *
 * Note:
 * If there is no such window in S that covers all characters in T, return the emtpy string "".
 *
 * If there are multiple such windows, you are guaranteed that there will always be only one unique minimum window in S.
 */
public class MinimumWindowSubstring {
    public String minWindow(String S, String T) {
        if(S == null || T == null || S.length() < T.length()){
            return "";
        }

        Map<Character, Integer> map = new HashMap<Character, Integer>();
        //initial map
        for(int i = 0; i < T.length(); i++){
            char c = T.charAt(i);
            if(map.containsKey(c)){
                map.put(c, map.get(c) + 1);
            } else {
                map.put(c, 1);
            }
        }
        int left = 0, minWin = S.length() + 1, minStart = 0, curLen = 0;
        for(int i = 0; i < S.length(); i++){
            char c = S.charAt(i);
            if(!map.containsKey(c)){
                continue;
            }
            map.put(c, map.get(c) - 1);

            //curLen is used to count how many valid characters in T are covered
            if(map.get(c) >= 0){
                curLen++;
            }
            //find a valid window
            while(curLen >= T.length()){
                //get window length and start
                if(i - left + 1 < minWin){
                    minWin = i - left + 1;
                    minStart = left;
                }
                //move the left of window
                char c1 = S.charAt(left);
                if(map.containsKey(c1)){
                    map.put(c1, map.get(c1) + 1);
                    //check if the window loose a valid character
                    if(map.get(c1) > 0){
                        curLen--;
                    }
                }
                left++;
            }
        }
        return minWin > S.length() ? "" : S.substring(minStart, minStart + minWin);
    }

    public static void main(String[] args) {
        MinimumWindowSubstring minimumWindowSubstring = new MinimumWindowSubstring();
        System.out.println(minimumWindowSubstring.minWindow("a", "a"));
    }
}

File: python_module/examples/Minimum_Window_Substring.py
from collections import Counter, defaultdict


class Solution:
    """
        Given two strings s and t of lengths m and n respectively, return the minimum window substring of s such that every character in t (including duplicates) is included in the window. If there is no such substring, return the empty string "".

        The testcases will be generated such that the answer is unique.

        A substring is a contiguous sequence of characters within the string.

        Example 1:

        Input: s = "ADOBECODEBANC", t = "ABC"
        Output: "BANC"
        Explanation: The minimum window substring "BANC" includes 'A', 'B', and 'C' from string t.
        Example 2:

        Input: s = "a", t = "a"
        Output: "a"
        Explanation: The entire string s is the minimum window.
        Example 3:

        Input: s = "a", t = "aa"
        Output: ""
        Explanation: Both 'a's from t must be included in the window.
        Since the largest window of s only has one 'a', return empty string.
    """
    def minWindow(self, s: str, t: str) -> str:
        if not s or not t:
            return ""

        # Dictionary which keeps a count of all the unique characters in t.
        dictT = Counter(t)
        # Number of unique characters in t, which need to be present in the desired window.
        requiredLength = len(dictT)
        # left and right pointer
        l, r = 0, 0

        # formed is used to keep track of how many unique characters in t are present in the current window in its desired frequency.
        # e.g. if t is "AABC" then the window must have two A's, one B and one C. Thus formed would be = 3 when all these conditions are met.
        formed = 0

        # Dictionary which keeps a count of all the unique characters in the current window.
        dictWindow = defaultdict(lambda : 0)
        ans = (float('inf'), None, None)
        while r < len(s):
            # Add one character from the right to the window
            character = s[r]
            dictWindow[character] += 1
            # If the frequency of the current character added equals to the desired count in t then increment the formed count by 1.
            if character in dictT and dictWindow[character] == dictT[character]:
                formed += 1

            # Try and contract the window till the point where it ceases to be 'desirable'.
            while l <= r and formed == requiredLength:
                character = s[l]
                # Save the smallest window until now.
                if r - l + 1 < ans[0]:
                    ans = (r - l + 1, l, r)

                # The character at the position pointed by the `left` pointer is no longer a part of the window.
                dictWindow[character] -= 1
                if character in dictT and dictWindow[character] < dictT[character]:
                    formed -= 1
                # Move the left pointer ahead, this would help to look for a new window.
                l += 1
            r += 1
        return "" if ans[0] == float("inf") else s[ans[1] : ans[2] + 1]




